### ISSUE TITLE
Implement einsum backprop rather than decomposing

### ISSUE THREAD

**Issue by pgmoka** [pytorch/pytorch#149133]
### 🚀 The feature, motivation and pitch

Currently when doing einsum backpropagation, the function decomposes into a series of IRs that achieves the einsum function. This is an issue for TPUs as the decomposition creates a significant performance impact due to the potential reshapes.

When looking at IRs for einsum at the moment, we will get something like:

```
IR {
  %0 = f32[] prim::Constant(), xla_shape=f32[]
  %1 = f32[3,3]{1,0} aten::expand(%0), xla_shape=f32[3,3]{1,0}
  %2 = f32[3,3,1]{2,1,0} aten::as_strided(%1), xla_shape=f32[3,3,1]{2,1,0}
  %3 = f32[3,3,1]{2,1,0} aten::as_strided(%2), xla_shape=f32[3,3,1]{2,1,0}
  %4 = f32[1,3,3]{2,1,0} aten::view(%3), xla_shape=f32[1,3,3]{2,1,0}
  %5 = f32[] prim::Constant(), xla_shape=f32[]
  %6 = f32[3,3]{1,0} aten::expand(%5), xla_shape=f32[3,3]{1,0}
  %7 = f32[3,3,1]{2,1,0} aten::as_strided(%6), xla_shape=f32[3,3,1]{2,1,0}
  %8 = f32[3,3,1]{2,1,0} aten::as_strided(%7), xla_shape=f32[3,3,1]{2,1,0}
  %9 = f32[1,3,3]{2,1,0} aten::view(%8), xla_shape=f32[1,3,3]{2,1,0}
  %10 = f32[1,3,3]{2,1,0} aten::matmul(%9, %4), xla_shape=f32[1,3,3]{2,1,0}
  %11 = f32[3,1,3]{2,1,0} aten::view(%10), xla_shape=f32[3,1,3]{2,1,0}
  %12 = f32[3,3,1]{2,1,0} aten::as_strided(%11), xla_shape=f32[3,3,1]{2,1,0}
  %13 = f32[3,3]{1,0} aten::view(%12), xla_shape=f32[3,3]{1,0}, ROOT=0
}
```
rather than a call to something like `aten::einsum` which can perform the einsum function more efficiently.

### Additional context

The lack of a backprop implementation from PyTorch caused an issue in PyTorchXLA (https://github.com/pytorch/xla/issues/8713). While we are able to create dispatch calls that resolve this issue, it creates potentially unknown edge cases, and it makes us acquire tech debt slowly over time.

**Comment by albanD** [comment]
@pgmoka do you have any reference that describes what the backward for a given einsum should be? I don't expect this is going to be trivial to derive.

**Comment by nikitaved** [comment]
But these are view operations. Do they still cause a lot of overhead? Currently, the contractions spanning several inputs are reduced to binary contractions. So, one possible improvement could be merging them -- but that still happens on the level of decompositions.

Of course, since einsum, in a nutshell, is a composition of elemwise-product + dim reductions + diagonalization, it is possible to write a naive backward which will be a variant of einsum for each input with some shared compute. The binary contraction order business optimizes performance in the forward, but I am not convinced that this order implies better performance in the backward, so, I agree, there is space for investigations here...
